### ISSUE TITLE
ENH: parallel processing for ufuncs

### ISSUE THREAD

**Issue by mattharrigan** [numpy/numpy#8208]
ufuncs support lots of great features, but one which seems to be missing is parallel processing.  Below is a quick benchmark which shows a ~3x reduction in time using 4 threads (my machine has 4 physical cores).  I don't know what the API should be, perhaps a ` np.parallel(True)` call would turn it on for subsequent calls as I would hate to have to litter the code with additional kwargs.  Also there would probably have to be some logic about the relative CPU intensity vs memory bandwith of the ufunc call since simple ones like addition probably wont go any faster and might be worse.  I read about python not really supporting multithreading, but will all the processing intensive numpy calls releasing the GIL I think it works just fine here.

```
import timeit

setup = """
import numpy as np
from concurrent.futures import ThreadPoolExecutor
size = 1e7
threads = 4
x = np.linspace(0, 2*np.pi, size)
"""

stmt = """
with ThreadPoolExecutor(max_workers=threads) as executor:
    indices = np.linspace(0, size, threads+1).astype(np.int)
    for i in range(threads):
        executor.submit(lambda x: np.sin(x, out=x), x[indices[i]:indices[i+1]])
"""

print(timeit.repeat(stmt, setup, number=10))
print(timeit.repeat("np.sin(x, out=x)", setup, number=10))
```

**Comment by shoyer** [comment]
[dask.array](http://dask.pydata.org/en/latest/array-overview.html) is a pretty solid way to do parallel ufuncs with NumPy and multi-threading.

First class support for parallel processing is probably out of scope for NumPy.

**Comment by rgommers** [comment]
> First class support for parallel processing is probably out of scope for NumPy.

Agreed, that's definitely out of scope. We should make it possible (releasing GIL when possible etc.), but we shouldn't add functions like ``np.parallel``.

**Comment by rgommers** [comment]
Besides ``dask.array`` there's also IPython parallel, ``joblib.Parallel``, and a bunch more (popular topic).

Going to close this, since it's out of scope. Thanks for the suggestion @mattharrigan

**Comment by mattharrigan** [comment]
I respect that its out of scope.  I think that in the long run high performance single threaded computing is a contradiction in terms.  I guess your view is that numpy is a building block towards that, not the complete solution.  Processing ufuncs in parallel is a trivially parallelizable task, but that's another contradiction in terms.

**Comment by rgommers** [comment]
>  I guess your view is that numpy is a building block towards that, not the complete solution.

Indeed.
